Engine Improvements and ClueBotImproved

rcswift opened the pull request:
I reorganized the relationship between 'bot' and 'board' a bit to open up some possibilities for future improvements.
- All players are referred to by absolute index now
- Added a 'listen(Turn)' method to the bot API so that bots are aware of other bots turns in case they want to track internal state. 
- Implemented a more robust information tracking concept in the board, based around 'CardInformation' using sets to track remaining possibilities for any cards in players' hands. This should be robust enough to deal with regular hanabi as well as duck variant.

rcswift committed:
Changed current_player to a @property

rcswift committed:
More @property decorators

rcswift committed:
Some reorganization of the bot/board relationship.

- Bots are created with a reference to the board.
- Board has been given a reset() method to be called after every game,
  like bots, so that the reference doesn't break.
- Bots are created with knowledge of their index.
- visible_hands still returns in turn-order, but it now returns a tuple
  with the player index so that all player indexing is absolute rather
  than relative.

rcswift committed:
Implemented a more robust system for tracking clues

rcswift committed:
Make better use of sets in the CardInformation class

rcswift committed:
Created ClueBotImproved which improves slightly on ClueBot

- Avoids cluing an already-clued card
- Picks color or number clue based on which touches fewer cards

c2huc2hu commented on bot.py:
```diff
@@ -4,35 +4,20 @@
-    def __init__(self):
-        pass
+    def __init__(self, board: Board, index: int):
+        self.board = board
```
I deliberately separated the `board` from the `bot` to make `bot.play` a pure function. This makes it easier to do hypotheticals and test a bot. Thoughts?

Although this means that stateful bots are discouraged (or need a `compute_state(board)` function)

rcswift replied:
I made this change with the idea that bots should eventually be stateful. 

With the updated information system board is aware of "Explicit" information (e.g. clue touched a card) and "Direct Implicit" information (e.g. clue did not touch a card), however "Indirect Implicit" information (e.g. play clue vs save clue, prompt, finesse) is much more difficult to capture in the board. 

If a bot wants to consider "Indirect Implicit" information, then it needs to keep track of who said what and when. That was the idea behind stateful bots and the 'Listen' method.

c2huc2hu replied:
With `self.turns`, it should be possible to read from that instead. Not sure if that's preferable though

rcswift replied:
I've gone back to the original pattern here. I agree with your point about testing hypotheticals.

However, `self.turns` is not enough of a history for the purposes of identifying play/save clues etc. because we don't store the starting deal so we can't re-create the state of the board purely from this history.

c2huc2hu commented on hanabi.py:
```diff
@@ -154,6 +155,22 @@ def _draw_card(self, index=None):
+    #
+    # Common helper functions
+    #
+    def is_playable(self, card: Card) -> bool:
```
This makes way more sense ty

c2huc2hu commented on hanabi.py:
```diff
@@ -27,10 +34,19 @@ class Play:
+        return f"{self.color}{self.number}"
+
+@dataclass
+class CardInformation:
```
I think it would be easier to simply merge the `CardInformation` and `Card` classes. This would be less bug-prone because rearranging cards becomes much simpler

c2huc2hu replied:
Also, you'd have to rename `color` -> `color_hints`, which I think would be a good for clarity anyways

c2huc2hu replied:
Should probably add a comment here that duck variant bots cannot access color or number attributes.

rcswift replied:
I've added the 'Variant' classes to deal with this now. When playing `VariantDuck` the number/color information does not update, only the `clued` flag.

c2huc2hu commented on hanabi.py:
```diff
@@ -62,7 +79,12 @@ def reset(self, seed=None, starting_player=0):
+
+    @property
+    def cards_per_player(self):
+        return 5 if self.num_players < 3 else 4
```
Rename to initial_cards_per_player, because towards the end of a game, a player will have fewer than this many cards

c2huc2hu commented on hanabi.py:
```diff
@@ -144,6 +133,79 @@ def evaluate(self, turn: Turn):
         self.turn_index += 1
 
+    def _clue_turn(self, turn: Clue):
+        target = turn.target
```
Far more sensible ty. Optional: maybe replace the argument name with `clue: Clue` instead

c2huc2hu commented on bot.py:
```diff
@@ -104,3 +104,37 @@ def play(self) -> Turn:
             return Clue(target=self.board.relative_player(1), color="r") # throwaway clue. this is technically not allowed
         return Discard(self.board.my_hand_size - 1)
+
+class ClueBotImproved(BaseBot):
```
Add your results to the README!

c2huc2hu commented on hanabi.py:
```diff
@@ -119,44 +132,117 @@ def evaluate(self, turn: Turn):
+                    self.information[target][idx].number.add(turn.number)
+                    self.information[target][idx].clued = True
+                else:
+                    self.information[target][idx].number.discard(turn.number)
```
Unless I'm missing something, the method for removing from a set should be `.remove` not `discard`?

rcswift replied:
.discard() succeeds if the element isn't present as opposed to .remove() which throws an exception.

c2huc2hu reviewed (changes requested):
See comments

rcswift committed:
Fixed some bugs with CardInformation and added some new helper functions

rcswift committed:
Created ClueBotMk3 which tries to do some cleverer things.

Average score has improved to ~14

rcswift committed:
Refinements to the 'Hand' and 'Information' API to make them similar.

Updated bots to use latest API

rcswift committed:
Added my bots to the readme

rcswift committed:
Created ListenerBot which tries to keep track of information on its own.

rcswift committed:
Changed updated information from INFO to DEBUG

rcswift committed:
Added 'cards_touched()' helper function

rcswift committed:
Created ListenerBotMk2 which improves on the clue rules in ListenerBot

rcswift committed:
Reversed changes to the board/bot relationship.

Bots no longer hold a pointer to the board, the board is passed as an
argument to play().

Updated all bots except ListenerBot and ListenerBotMk2 to work with this
again. ListenerBots need to be re-worked some more to follow this paradigm.

rcswift committed:
Created a Variant class to handle implementing variations on the rules

(i.e. Normal, Duck, Rainbow, etc.)

rcswift committed:
Removed ListenerBot and ListenerBotMk2 and replaced them with
ClueBotAdvanced which uses the same rules and gets the same score.

rcswift committed:
Created LookaheadBot
